### ISSUE TITLE
Non-deterministic code generated by derive_more macro

### ISSUE THREAD

**Issue by edwin0cheng** [JelteF/derive_more#124]
Hi, I am one of the contributors of rust-analyzer who are working on proc-macro support. We found that in `derive_more`, there are some [usages][1] of  `std::collection::Hashmap` , such that the generated code is non-deterministic. (The order of generated tokens are inconsistent)

Although in general it does not affect end-user experience, but derive-more should still be fixed to be deterministic. Otherwise crates using it have non reproducible builds. And replacing `Hashmap` to [`FxHashmap`][fx-hash] will be enough to solve this bug I think. ( I am happy to submit a PR for that if it is acceptable)

Here is the related issue in rust-analzyer:  https://github.com/rust-analyzer/rust-analyzer/issues/4315#issuecomment-624606306

[1]: https://github.com/JelteF/derive_more/search?q=Hashmap&unscoped_q=Hashmap
[fx-hash]: https://github.com/cbreeden/fxhash

**Comment by bjorn3** [comment]
Using `BTreeMap` would also work if the key implements `Ord`.

**Comment by JelteF** [comment]
A PR to fix this is appreciated. BTreeMap would be preferred, since that is already in the standard library.

**Comment by edwin0cheng** [comment]
Some keys (e.g `syn::Type`) does not implements `Ord`, so I think I have to use `FxHashmap` , but [`rustc-hash::FxHashmap`][1] basically is just a single file implementation, If you prefer not adding other deps, I could copy and paste it. Which way would you prefer ?

[1]: https://github.com/rust-lang/rustc-hash/blob/master/src/lib.rs

**Comment by JelteF** [comment]
I'd rather add ordering to the hashmap then using increnenting integer to
keep track of which key was inserted first.

**Comment by edwin0cheng** [comment]
For `from` trait, it is easy, but [`display`][1] trait is quite a big change, so it is okay ? Or do you mean you want like to define a new struct , some like :
```rust
struct OrderedHashmap {
...
}
```
and replaced the `std::collection::Hashmap`?

[1]: https://github.com/JelteF/derive_more/blob/0f899f1e29ac1c3ede0cbf1c41ad50c3bf23f680/src/display.rs

**Comment by CreepySkeleton** [comment]
> I'd rather add ordering to the hashmap then using increnenting integer to
keep track of which key was inserted first.

Actually, instead of reinventing the wheel, I would recommend [`indexmap](https://docs.rs/indexmap/1.3.2/indexmap/) for that.

**Comment by JelteF** [comment]
I was thinking more using:
Instead of `Hashmap<K, V>` have `Hashmap<K, (i32, V)>` and whenever you insert you can do something like:
```rust
hmap.or_insert_with(||(hmap.values().map(|x| x.0).max() + 1, defaultvalue))
```
If the current max can be easily stored somewhere, like in from that's preferable. But I wouldn't make this an overengineered solution. 
I think indexmap looks very cool, but it has a lot of code and so it seems a bit more than needed.
FxHashMap is indeed very small, but it's unclear to me how it actually solves the problem.

**Comment by edwin0cheng** [comment]
> FxHashMap is indeed very small, but it's unclear to me how it actually solves the problem.

The problem of `std::collection::Hashmap` is, it is designed to resistance against HashDoS attacks such that it is randomly seeded and non-deterministic intentionally. OTHO, `FxHashMap` is designed to be deterministic and non-security focused such that it is idempotent. (No guarantee on the orders of key pair, but  guaranteed with the same orders for the same input)

**Comment by edwin0cheng** [comment]
And in fact `std::collection::HashSet` had the same problem too. So for simplicity reason, I would  recommend drop-in replace container solution rather than insertion key one.

**Comment by JelteF** [comment]
Googling around a bit more you should get deterministic hashing by using
DefaultHasher::new() as the hasher everywhere:
https://doc.rust-lang.org/std/collections/hash_map/struct.DefaultHasher.html

The RandomState hasher apparently is the source of the non determinism.

This seems like the simplest and cleanest solution to me.

**Comment by JelteF** [comment]
So using the following everywhere should fix it:

HashMap::with_hasher(DefaultHasher::new())

**Comment by edwin0cheng** [comment]
Um.. `DefaultHasher` is used inside `RandomState` , the actual trait is `BuildHasher` which only implemented by `RandomState`:

here is the playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=abbb1677ac75958202cc62c121d36f35

[1]: https://doc.rust-lang.org/std/hash/trait.BuildHasher.html

**Comment by JelteF** [comment]
Okay, so slightly more involved. You need to create a type that implements
BuildHasher and always returns DefaultHasher::new() from its build_hasher
method. Then this type should be given to with_hasher.

RandomState adds the randomness to DefaultHasher
https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#2445

DefaultHasher itself is deterministic as long as its instantiated with the
same values, which new/default does.

You could possibly even make a type alias of this deterministic hashmap to
make it easy to use everywhere: Hashmap<K, V, DeterministicState>

**Comment by JelteF** [comment]
The fix for this was released as v0.99.6 a few days ago. @edwin0cheng can you confirm that did this indeed fixes the issue with rust-analyzer?

**Comment by edwin0cheng** [comment]
It is an general issue for proc-macro support in RA (non-deterministic proc-macro), so we already "fixed" the issue on RA's side by making sure that we never GC expansions from proc-macros.

But I reverted that "fixes" and test this fixes  (By `cargo update`), and it works ! 

Thanks !
